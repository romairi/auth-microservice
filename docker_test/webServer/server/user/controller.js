import axios from "axios";
import { StatusCodes } from "http-status-codes";
import { LOGIN_API } from "./constants";

export async function checkAuth(req, res, next) {
  console.log("checkAuth", req.body);

  res.json({
    status: StatusCodes.OK,
  });
}

export async function login(req, res, next) {
  console.log(LOGIN_API);
  try {
    const response = await axios.post(LOGIN_API, {
      data: req.body.data,
    });
    console.log(response);
  } catch (err) {
    console.log(err.response.data);
  }

  res.json({
    status: StatusCodes.OK,
  });
}

export async function register(req, res, next) {
  console.log("register");

  res.json({
    status: StatusCodes.OK,
  });
}
