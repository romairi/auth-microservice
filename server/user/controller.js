import {serverConfig} from "../config/serverConfig";
import {StatusCodes} from 'http-status-codes';
import UserModel from './model';
import bcrypt from 'bcryptjs'
import jwt from 'jsonwebtoken';
import {
    TOKEN_EXPIRATION_TIME,
    COOKIE_EXPIRATION_TIME,
    ERROR_MESSAGE,
    ERROR_EMAIL_EXIST_MESSAGE
} from "./constants";


export async function login(req, res, next) {
    const {email, password} = req.body;
    try {
        const user = await UserModel.findOne({email});
        if (!user) {
            res.status(StatusCodes.BAD_REQUEST).json({type: 'error', message: ERROR_MESSAGE});
            return;
        }
        const isMatch = await bcrypt.compare(password, user.password);

        if (!isMatch) {
            return res.status(StatusCodes.BAD_REQUEST).json({type: 'error', message: ERROR_MESSAGE});
        }

        return loadUser(res, user);

    } catch (err) {
        next(err);
    }
}

export async function register(req, res, next) {
    const {username, email, password} = req.body;

    try {
        const foundUser = await UserModel.findOne({email});
        if (foundUser) {
            res.status(StatusCodes.BAD_REQUEST).json({type: 'error', message: ERROR_EMAIL_EXIST_MESSAGE});
            return;
        }
        // encryption of user password
        const salt = await bcrypt.genSalt(10);
        const hashedPassword = await bcrypt.hash(password, salt);

        // creating a new user
        const newUser = new UserModel({username, password: hashedPassword, email});
        const user = await newUser.save();

        return loadUser(res, user);
    } catch (err) {
        next(err);
    }
}

export function logout(req, res, next) {
    if (req.authenticated) {
        res.clearCookie('token');
    }
    res.send('logout');
}

export function loadUser(res, user) {
    const token = jwt.sign({_id: user._id}, serverConfig.jwt.secret, {expiresIn: TOKEN_EXPIRATION_TIME});
    const {password: userPassword, ...userArgs} = user.toObject();

    res
        .cookie('token', token, {
            expires: new Date(Date.now() + COOKIE_EXPIRATION_TIME),
            secure: false,
            httpOnly: true
        })
        .json(userArgs);
    return res;
}
