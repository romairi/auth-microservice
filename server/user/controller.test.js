import jwt from "jsonwebtoken";

import { loadUser } from "./controller";

jest.mock("jsonwebtoken", () => ({
  sign: jest.fn(() => "TOKEN"),
}));

jest.mock("../config/serverConfig", () => ({
  serverConfig: { jwt: { secret: "SECRET" } },
}));

describe("controller", () => {
  describe("loadUser", () => {
    const userMock = {
      _id: 1,
      toObject: () => ({ password: "PASSWORD", userName: "USER_NAME" }),
    };

    const resMock = {
      json: jest.fn(),
    };

    it("should load the user", () => {
      const result = loadUser(resMock, userMock);

      expect(resMock.json.mock.calls.pop()).toMatchSnapshot();
      expect(jwt.sign.mock.calls.pop()).toMatchSnapshot();
      expect(result).toBe(resMock);
    });
  });
});
