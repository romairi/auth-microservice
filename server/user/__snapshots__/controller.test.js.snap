// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller loadUser should load the user 1`] = `
Array [
  Object {
    "token": "TOKEN",
    "user": Object {
      "userName": "USER_NAME",
    },
  },
]
`;

exports[`controller loadUser should load the user 2`] = `
Array [
  Object {
    "_id": 1,
  },
  "SECRET",
  Object {
    "expiresIn": "24h",
  },
]
`;
